code/src/main/kotlin/com/expediagroup/sdk/lodgingconnectivity/graphql/model/paging/PaginatedStream.kt

```diff
@@ -0,0 +1,21 @@
+package com.expediagroup.sdk.lodgingconnectivity.graphql.model.paging
+
+import java.util.stream.Stream
+import kotlin.streams.asStream
+
+abstract class PaginatedStream<T> {
+    protected var page: List<T?> = emptyList()
+    protected var index = 0
+
+    fun stream(): Stream<T> = generateSequence { nextItem() }.asStream()
+
+    protected abstract fun nextItem(): T?
+
+    protected fun fetchNextPage(cb: () -> List<T?>) {
+        page = cb()
+    }
+
+    protected fun resetIndex() {
+        index = 0
+    }
+}
```

Comment:
Consider using a Queue (ArrayDeque) instead of List + index here. It would make the flow cleaner and remove the need for index tracking.
Plus, I would go with a more descriptive parameter name instead of `cb`.
Something like:

```suggestion
abstract class PaginatedStream<T> {
    private var currentPage: Queue<T?> = ArrayDeque()
    
    fun stream(): Stream<T> = generateSequence { nextItem() }.asStream()
    
    protected abstract fun nextItem(): T?
    
    protected fun fetchNextPage(pageSupplier: () -> List<T?>) {
        currentPage = ArrayDeque(pageSupplier())
    }
    
    protected fun peekCurrentPage(): T? = currentPage.peek()
    
    protected fun pollCurrentPage(): T? = currentPage.poll()
    
    protected fun isCurrentPageEmpty(): Boolean = currentPage.isEmpty()
}
```

Reply:
Nice 🚀 - Done!